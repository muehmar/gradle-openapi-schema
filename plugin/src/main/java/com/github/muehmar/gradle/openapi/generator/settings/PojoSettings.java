package com.github.muehmar.gradle.openapi.generator.settings;

import static com.github.muehmar.gradle.openapi.util.Booleans.not;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.model.name.Name;
import com.github.muehmar.gradle.openapi.task.TaskIdentifier;
import com.github.muehmar.gradle.openapi.warnings.Warning;
import com.github.muehmar.gradle.openapi.warnings.WarningsContext;
import io.github.muehmar.pojobuilder.annotations.FieldBuilder;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import java.io.Serializable;
import java.util.List;
import lombok.Value;
import lombok.With;

@Value
@With
@PojoBuilder
public class PojoSettings implements Serializable {
  JsonSupport jsonSupport;
  String packageName;
  String suffix;
  StagedBuilderSettings stagedBuilder;
  String builderMethodPrefix;
  boolean enableValidation;
  boolean nonStrictOneOfValidation;

  ValidationApi validationApi;
  List<ClassTypeMapping> classTypeMappings;
  List<FormatTypeMapping> formatTypeMappings;
  EnumDescriptionSettings enumDescriptionSettings;
  GetterSuffixes getterSuffixes;
  ValidationMethods validationMethods;
  List<String> excludeSchemas;
  PojoNameMappings pojoNameMappings;

  TaskIdentifier taskIdentifier;

  public boolean isJacksonJson() {
    return jsonSupport.equals(JsonSupport.JACKSON);
  }

  public boolean isEnableValidation() {
    return enableValidation;
  }

  public boolean isNonStrictOneOfValidation() {
    return nonStrictOneOfValidation;
  }

  public List<ClassTypeMapping> getClassTypeMappingsList() {
    return classTypeMappings;
  }

  public List<FormatTypeMapping> getFormatTypeMappingsList() {
    return formatTypeMappings;
  }

  public List<String> getExcludeSchemasList() {
    return excludeSchemas;
  }

  public PList<ClassTypeMapping> getClassTypeMappings() {
    return PList.fromIter(classTypeMappings);
  }

  public PList<FormatTypeMapping> getFormatTypeMappings() {
    return PList.fromIter(formatTypeMappings);
  }

  public ExcludedSchemas getExcludedSchemas() {
    return ExcludedSchemas.fromExcludedPojoNames(
        PList.fromIter(excludeSchemas).map(Name::ofString));
  }

  public boolean isEnableStagedBuilder() {
    return stagedBuilder.isEnabled();
  }

  public boolean isDisableStagedBuilder() {
    return !isEnableStagedBuilder();
  }

  public StagedBuilderSettings getStagedBuilder() {
    return stagedBuilder;
  }

  @FieldBuilder(fieldName = "classTypeMappings")
  public static List<ClassTypeMapping> classTypeMappings(
      PList<ClassTypeMapping> classTypeMappings) {
    return classTypeMappings.toArrayList();
  }

  @FieldBuilder(fieldName = "formatTypeMappings")
  public static List<FormatTypeMapping> formatTypeMappings(
      PList<FormatTypeMapping> formatTypeMappings) {
    return formatTypeMappings.toArrayList();
  }

  public TypeMappings getTypeMappings() {
    return new TypeMappings(PList.fromIter(classTypeMappings), PList.fromIter(formatTypeMappings));
  }

  public PojoNameMapping pojoNameMapping() {
    return PList.fromIter(pojoNameMappings.getConstantNameMappings())
        .map(
            constantNameMapping ->
                PojoNameMapping.replaceConstant(
                    constantNameMapping.getConstant(), constantNameMapping.getReplacement()))
        .foldLeft(PojoNameMapping.noMapping(), PojoNameMapping::andThen);
  }

  public void validate() {
    classTypeMappings.forEach(
        classTypeMapping -> {
          if (not(classTypeMapping.getTypeConversion().isPresent())) {
            final Warning warning = Warning.missingMappingConversion(classTypeMapping);
            WarningsContext.addWarningForTask(taskIdentifier, warning);
          }
        });

    formatTypeMappings.forEach(
        formatTypeMapping -> {
          if (not(formatTypeMapping.getTypeConversion().isPresent())) {
            final Warning warning = Warning.missingMappingConversion(formatTypeMapping);
            WarningsContext.addWarningForTask(taskIdentifier, warning);
          }
        });
  }
}
