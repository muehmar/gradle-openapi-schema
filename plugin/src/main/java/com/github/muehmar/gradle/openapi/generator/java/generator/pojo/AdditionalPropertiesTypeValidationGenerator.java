package com.github.muehmar.gradle.openapi.generator.java.generator.pojo;

import static com.github.muehmar.gradle.openapi.generator.java.generator.shared.ValidationGenerator.assertTrue;
import static com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties.additionalPropertiesName;
import static io.github.muehmar.codegenerator.Generator.constant;

import com.github.muehmar.gradle.openapi.generator.java.generator.shared.SettingsFunctions;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.jackson.JacksonAnnotationGenerator;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaObjectPojo;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.java.MethodGenBuilder;

public class AdditionalPropertiesTypeValidationGenerator {
  private AdditionalPropertiesTypeValidationGenerator() {}

  public static Generator<JavaObjectPojo, PojoSettings>
      additionalPropertiesTypeValidationGenerator() {
    return annotatedCorrectTypeMethod()
        .filter(JavaAdditionalProperties::isAllowed)
        .filter(JavaAdditionalProperties::isNotValueAnyType)
        .contraMap(JavaObjectPojo::getAdditionalProperties);
  }

  private static Generator<JavaAdditionalProperties, PojoSettings> annotatedCorrectTypeMethod() {
    return JacksonAnnotationGenerator.<JavaAdditionalProperties>jsonIgnore()
        .append(
            assertTrue(
                props ->
                    String.format(
                        "Not all additional properties are instances of %s",
                        props.getType().getQualifiedClassName().getClassName())))
        .append(correctTypeMethod());
  }

  private static Generator<JavaAdditionalProperties, PojoSettings> correctTypeMethod() {
    return MethodGenBuilder.<JavaAdditionalProperties, PojoSettings>create()
        .modifiers(SettingsFunctions::validationMethodModifiers)
        .noGenericTypes()
        .returnType("boolean")
        .methodName("isAllAdditionalPropertiesHaveCorrectType")
        .noArguments()
        .content(
            constant(
                String.format(
                    "return getAdditionalProperties().size() == %s.size();",
                    additionalPropertiesName())))
        .build();
  }
}
