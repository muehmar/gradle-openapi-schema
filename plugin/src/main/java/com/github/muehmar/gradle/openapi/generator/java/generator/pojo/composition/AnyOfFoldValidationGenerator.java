package com.github.muehmar.gradle.openapi.generator.java.generator.pojo.composition;

import static com.github.muehmar.gradle.openapi.generator.java.model.name.MethodNames.Composition.AnyOf.foldAnyOfMethodName;
import static com.github.muehmar.gradle.openapi.generator.java.model.name.MethodNames.Composition.AnyOf.getAnyOfMethodName;
import static com.github.muehmar.gradle.openapi.generator.java.model.name.MethodNames.Composition.AnyOf.getAnyOfValidCountMethodName;
import static io.github.muehmar.codegenerator.Generator.constant;

import ch.bluecare.commons.data.NonEmptyList;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.DeprecatedMethodGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.SettingsFunctions;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.jackson.JacksonAnnotationGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.validation.ValidationAnnotationGenerator;
import com.github.muehmar.gradle.openapi.generator.java.model.composition.JavaAnyOfComposition;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaObjectPojo;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.java.MethodGen;
import io.github.muehmar.codegenerator.java.MethodGenBuilder;

public class AnyOfFoldValidationGenerator {

  private AnyOfFoldValidationGenerator() {}

  public static Generator<JavaObjectPojo, PojoSettings> anyOfFoldValidationGenerator() {
    return Generator.<JavaObjectPojo, PojoSettings>emptyGen()
        .appendOptional(
            annotatedMethod(), p -> p.getAnyOfComposition().map(JavaAnyOfComposition::getPojos));
  }

  private static Generator<NonEmptyList<JavaObjectPojo>, PojoSettings> annotatedMethod() {
    final MethodGen<NonEmptyList<JavaObjectPojo>, PojoSettings> method =
        MethodGenBuilder.<NonEmptyList<JavaObjectPojo>, PojoSettings>create()
            .modifiers(SettingsFunctions::validationMethodModifiers)
            .noGenericTypes()
            .returnType("List<Object>")
            .methodName(getAnyOfMethodName().asString())
            .noArguments()
            .doesNotThrow()
            .content(methodContent())
            .build();

    return DeprecatedMethodGenerator
        .<NonEmptyList<JavaObjectPojo>>deprecatedJavaDocAndAnnotationForValidationMethod()
        .append(ValidationAnnotationGenerator.validAnnotation())
        .append(JacksonAnnotationGenerator.jsonIgnore())
        .append(method);
  }

  private static Generator<NonEmptyList<JavaObjectPojo>, PojoSettings> methodContent() {
    return Generator.<NonEmptyList<JavaObjectPojo>, PojoSettings>emptyGen()
        .append(w -> w.println("if (%s() == 0) {", getAnyOfValidCountMethodName()))
        .append(constant("return null;"), 1)
        .append(constant("}"))
        .append(
            (pojos, s, w) ->
                w.println(
                    "return %s(%s);",
                    foldAnyOfMethodName(),
                    pojos.map(name -> "dto -> dto").toPList().mkString(", ")));
  }
}
