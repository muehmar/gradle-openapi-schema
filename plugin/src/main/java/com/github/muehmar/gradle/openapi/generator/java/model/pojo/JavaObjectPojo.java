package com.github.muehmar.gradle.openapi.generator.java.model.pojo;

import static com.github.muehmar.gradle.openapi.generator.java.generator.pojo.MemberContentBuilder.fullMemberContentBuilder;
import static com.github.muehmar.gradle.openapi.generator.java.generator.pojo.WitherContentBuilder.fullWitherContentBuilder;
import static com.github.muehmar.gradle.openapi.generator.java.generator.shared.misc.ConstructorContentBuilder.fullConstructorContentBuilder;
import static com.github.muehmar.gradle.openapi.generator.java.generator.shared.misc.EqualsContentBuilder.fullEqualsContentBuilder;
import static com.github.muehmar.gradle.openapi.generator.java.generator.shared.misc.HashCodeContentBuilder.fullHashCodeContentBuilder;
import static com.github.muehmar.gradle.openapi.generator.java.generator.shared.misc.ToStringContentBuilder.fullToStringContentBuilder;
import static com.github.muehmar.gradle.openapi.util.Booleans.not;

import ch.bluecare.commons.data.NonEmptyList;
import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.exception.OpenApiGeneratorException;
import com.github.muehmar.gradle.openapi.generator.java.generator.pojo.MemberGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.pojo.WitherGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.misc.EqualsGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.misc.HashCodeGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.misc.PojoConstructorGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.misc.ToStringGenerator;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMember;
import com.github.muehmar.gradle.openapi.generator.java.model.PojoType;
import com.github.muehmar.gradle.openapi.generator.java.model.TechnicalPojoMember;
import com.github.muehmar.gradle.openapi.generator.java.model.composition.JavaAllOfComposition;
import com.github.muehmar.gradle.openapi.generator.java.model.composition.JavaAnyOfComposition;
import com.github.muehmar.gradle.openapi.generator.java.model.composition.JavaOneOfComposition;
import com.github.muehmar.gradle.openapi.generator.java.model.name.JavaName;
import com.github.muehmar.gradle.openapi.generator.java.model.name.JavaPojoName;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.auxiliaryy.AnyOfContainer;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.auxiliaryy.OneOfContainer;
import com.github.muehmar.gradle.openapi.generator.model.constraints.Constraints;
import com.github.muehmar.gradle.openapi.generator.model.name.SchemaName;
import com.github.muehmar.gradle.openapi.generator.model.pojo.ObjectPojo;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import io.github.muehmar.codegenerator.java.JavaModifier;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import java.util.Comparator;
import java.util.Optional;
import java.util.function.Function;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@PojoBuilder(packagePrivateBuilder = true)
@EqualsAndHashCode
@ToString
public class JavaObjectPojo implements JavaPojo {
  private final JavaPojoName name;
  private final SchemaName schemaName;
  private final String description;
  private final PList<JavaPojoMember> members;
  private final Optional<JavaAllOfComposition> allOfComposition;
  private final Optional<JavaOneOfComposition> oneOfComposition;
  private final Optional<JavaAnyOfComposition> anyOfComposition;
  private final PojoType type;
  private final PList<JavaRequiredAdditionalProperty> requiredAdditionalProperties;
  private final JavaAdditionalProperties additionalProperties;
  private final Constraints constraints;

  JavaObjectPojo(
      JavaPojoName name,
      SchemaName schemaName,
      String description,
      PList<JavaPojoMember> members,
      Optional<JavaAllOfComposition> allOfComposition,
      Optional<JavaOneOfComposition> oneOfComposition,
      Optional<JavaAnyOfComposition> anyOfComposition,
      PojoType type,
      PList<JavaRequiredAdditionalProperty> requiredAdditionalProperties,
      JavaAdditionalProperties additionalProperties,
      Constraints constraints) {
    this.name = name;
    this.schemaName = schemaName;
    this.description = Optional.ofNullable(description).orElse("");
    this.members = members;
    this.allOfComposition = allOfComposition;
    this.oneOfComposition = oneOfComposition;
    this.anyOfComposition = anyOfComposition;
    this.type = type;
    this.requiredAdditionalProperties = requiredAdditionalProperties;
    this.additionalProperties = additionalProperties;
    this.constraints = constraints;
    assertPropertiesHaveNotSameNameAndDifferentAttributes(
        name, members, allOfComposition, oneOfComposition, anyOfComposition);
  }

  private static void assertPropertiesHaveNotSameNameAndDifferentAttributes(
      JavaPojoName name,
      PList<JavaPojoMember> members,
      Optional<JavaAllOfComposition> allOfComposition,
      Optional<JavaOneOfComposition> oneOfComposition,
      Optional<JavaAnyOfComposition> anyOfComposition) {
    final PList<JavaPojoMember> allMembers =
        members
            .concat(allOfComposition.map(JavaAllOfComposition::getMembers).orElseGet(PList::empty))
            .concat(oneOfComposition.map(JavaOneOfComposition::getMembers).orElseGet(PList::empty))
            .concat(anyOfComposition.map(JavaAnyOfComposition::getMembers).orElseGet(PList::empty));

    final PList<JavaPojoMember> invalidMembers =
        allMembers
            .filter(
                member1 ->
                    allMembers.exists(member2 -> sameNameButDifferentAttributes(member1, member2)))
            .duplicates(Comparator.comparing(m -> m.getName().asString()));

    if (invalidMembers.nonEmpty()) {
      throw new OpenApiGeneratorException(
          String.format(
              "Cannot create DTO %s: Two or more properties (defined either in properties or in one of the schemas used "
                  + "for a composition (allOf, oneOf or anyOf) have the same name but different attributes (e.g. type, "
                  + "constraints)! Invalid properties: [%s].",
              name, invalidMembers.map(JavaPojoMember::getName).mkString(", ")));
    }
  }

  private static boolean sameNameButDifferentAttributes(JavaPojoMember m1, JavaPojoMember m2) {
    return m1.getName().equals(m2.getName())
        && not(m1.asObjectMember().equals(m2.asObjectMember()));
  }

  public static JavaPojoWrapResult wrap(ObjectPojo objectPojo, TypeMappings typeMappings) {
    if (objectPojo.containsNoneDefaultPropertyScope()) {
      return JavaPojoWrapResultBuilder.create()
          .defaultPojo(createForType(objectPojo, typeMappings, PojoType.DEFAULT))
          .andAllOptionals()
          .requestPojo(createForType(objectPojo, typeMappings, PojoType.REQUEST))
          .responsePojo(createForType(objectPojo, typeMappings, PojoType.RESPONSE))
          .build();
    } else {
      return JavaPojoWrapResult.ofDefaultPojo(
          createForType(objectPojo, typeMappings, PojoType.DEFAULT));
    }
  }

  private static JavaObjectPojo createForType(
      ObjectPojo objectPojo, TypeMappings typeMappings, PojoType type) {
    final PList<JavaPojoMember> members =
        objectPojo
            .getMembers()
            .filter(member -> type.includesPropertyScope(member.getPropertyScope()))
            .map(member -> JavaPojoMember.wrap(member, typeMappings));
    final JavaAdditionalProperties javaAdditionalProperties =
        JavaAdditionalProperties.wrap(objectPojo.getAdditionalProperties(), typeMappings);
    return new JavaObjectPojo(
        JavaPojoName.fromPojoName(type.mapName(objectPojo.getName().getPojoName())),
        objectPojo.getName().getSchemaName(),
        objectPojo.getDescription(),
        members,
        objectPojo
            .getAllOfComposition()
            .map(comp -> JavaAllOfComposition.wrap(comp, type, typeMappings)),
        objectPojo
            .getOneOfComposition()
            .map(
                comp ->
                    JavaOneOfComposition.wrap(
                        comp, objectPojo.getDiscriminator(), type, typeMappings)),
        objectPojo
            .getAnyOfComposition()
            .map(comp -> JavaAnyOfComposition.wrap(comp, type, typeMappings)),
        type,
        objectPojo
            .getRequiredAdditionalProperties()
            .map(
                propName ->
                    JavaRequiredAdditionalProperty.fromNameAndType(
                        propName, javaAdditionalProperties.getType())),
        javaAdditionalProperties,
        objectPojo.getConstraints());
  }

  @Override
  public JavaName getSchemaName() {
    return JavaName.fromName(schemaName.asName());
  }

  @Override
  public JavaName getClassName() {
    return name.asJavaName();
  }

  public JavaPojoName getJavaPojoName() {
    return name;
  }

  public JavaName prefixedClassNameForMethod(String prefix) {
    return name.asJavaName().prefixedMethodName(prefix);
  }

  @Override
  public String getDescription() {
    return description;
  }

  @Override
  public PojoType getType() {
    return type;
  }

  public boolean isSimpleMapPojo() {
    return members.isEmpty()
        && not(allOfComposition.isPresent())
        && not(oneOfComposition.isPresent())
        && not(anyOfComposition.isPresent());
  }

  public boolean hasRequiredMembers() {
    return members.exists(JavaPojoMember::isRequired);
  }

  public boolean hasNotRequiredMembers() {
    return not(hasRequiredMembers());
  }

  public PList<JavaPojoMember> getMembers() {
    return members;
  }

  /** This includes also any possible required additional property. */
  public int getRequiredMemberCount() {
    return members.filter(JavaPojoMember::isRequired).size() + requiredAdditionalProperties.size();
  }

  public PList<JavaPojoMember> getAllMembersForComposition() {
    return getMembers()
        .map(member -> member.asInnerEnumOf(getClassName()))
        .concat(getComposedMembers());
  }

  public PList<JavaPojoMember> getComposedMembers() {
    final PList<JavaPojoMember> allOfMembers =
        allOfComposition.map(JavaAllOfComposition::getMembers).orElseGet(PList::empty);
    final PList<JavaPojoMember> oneOfMembers =
        oneOfComposition.map(JavaOneOfComposition::getMembers).orElseGet(PList::empty);
    final PList<JavaPojoMember> anyOfMembers =
        anyOfComposition.map(JavaAnyOfComposition::getMembers).orElseGet(PList::empty);
    return allOfMembers.concat(oneOfMembers).concat(anyOfMembers).distinct(Function.identity());
  }

  public PList<JavaPojoMember> getAllMembers() {
    return members.concat(getComposedMembers()).distinct(Function.identity());
  }

  public PList<TechnicalPojoMember> getTechnicalMembers() {
    return getAllMembers()
        .flatMap(JavaPojoMember::getTechnicalMembers)
        .add(additionalProperties.asTechnicalPojoMember());
  }

  public Optional<JavaAllOfComposition> getAllOfComposition() {
    return allOfComposition;
  }

  public Optional<JavaOneOfComposition> getOneOfComposition() {
    return oneOfComposition;
  }

  public PList<JavaObjectPojo> getOneOfPojos() {
    return oneOfComposition
        .map(JavaOneOfComposition::getPojos)
        .map(NonEmptyList::toPList)
        .orElseGet(PList::empty);
  }

  public boolean hasOneOfComposition() {
    return oneOfComposition.isPresent();
  }

  public Optional<JavaAnyOfComposition> getAnyOfComposition() {
    return anyOfComposition;
  }

  public PList<JavaObjectPojo> getAnyOfPojos() {
    return anyOfComposition
        .map(JavaAnyOfComposition::getPojos)
        .map(NonEmptyList::toPList)
        .orElseGet(PList::empty);
  }

  public boolean hasAnyOfComposition() {
    return anyOfComposition.isPresent();
  }

  public PList<JavaRequiredAdditionalProperty> getRequiredAdditionalProperties() {
    return requiredAdditionalProperties;
  }

  public JavaAdditionalProperties getAdditionalProperties() {
    return additionalProperties;
  }

  public Constraints getConstraints() {
    return constraints;
  }

  public MemberGenerator.MemberContent getMemberContent() {
    return fullMemberContentBuilder()
        .isArrayPojo(false)
        .members(getAllMembers().flatMap(JavaPojoMember::getTechnicalMembers))
        .additionalProperties(additionalProperties)
        .build();
  }

  public HashCodeGenerator.HashCodeContent getHashCodeContent() {
    return fullHashCodeContentBuilder().technicalPojoMembers(getTechnicalMembers()).build();
  }

  public EqualsGenerator.EqualsContent getEqualsContent() {
    return fullEqualsContentBuilder()
        .className(getClassName())
        .technicalPojoMembers(getTechnicalMembers())
        .build();
  }

  public ToStringGenerator.ToStringContent getToStringContent() {
    return fullToStringContentBuilder()
        .className(getClassName())
        .technicalPojoMembers(getTechnicalMembers())
        .build();
  }

  public PojoConstructorGenerator.ConstructorContent getConstructorContent() {
    return fullConstructorContentBuilder()
        .isArray(false)
        .className(getClassName())
        .members(getAllMembers().flatMap(JavaPojoMember::getTechnicalMembers))
        .modifier(Optional.of(JavaModifier.PUBLIC).filter(ignore -> not(isSimpleMapPojo())))
        .additionalProperties(additionalProperties)
        .build();
  }

  public WitherGenerator.WitherContent getWitherContent() {
    return fullWitherContentBuilder()
        .className(getClassName())
        .membersForWithers(getMembers())
        .technicalPojoMembers(getTechnicalMembers())
        .build();
  }

  public Optional<OneOfContainer> getOneOfContainer() {
    return oneOfComposition.map(composition -> new OneOfContainer(name, composition));
  }

  public Optional<AnyOfContainer> getAnyOfContainer() {
    return anyOfComposition.map(composition -> new AnyOfContainer(name, composition));
  }

  @Override
  public <T> T fold(
      Function<JavaArrayPojo, T> onArrayPojo,
      Function<JavaEnumPojo, T> onEnumPojo,
      Function<JavaObjectPojo, T> onObjectPojo) {
    return onObjectPojo.apply(this);
  }
}
