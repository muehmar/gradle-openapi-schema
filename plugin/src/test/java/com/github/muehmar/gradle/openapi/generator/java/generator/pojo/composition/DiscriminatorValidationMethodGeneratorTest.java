package com.github.muehmar.gradle.openapi.generator.java.generator.pojo.composition;

import static com.github.muehmar.gradle.openapi.generator.java.generator.pojo.composition.DiscriminatorValidationMethodGenerator.discriminatorValidationMethodGenerator;
import static com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojos.illegalIdentifierPojo;
import static com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojos.sampleObjectPojo1;
import static com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojos.sampleObjectPojo2;
import static com.github.muehmar.gradle.openapi.generator.settings.TestPojoSettings.defaultTestSettings;
import static com.github.muehmar.gradle.openapi.snapshot.SnapshotUtil.writerSnapshot;
import static io.github.muehmar.codegenerator.writer.Writer.javaWriter;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import ch.bluecare.commons.data.NonEmptyList;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMembers;
import com.github.muehmar.gradle.openapi.generator.java.model.composition.JavaOneOfCompositions;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaObjectPojo;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojos;
import com.github.muehmar.gradle.openapi.generator.model.Discriminator;
import com.github.muehmar.gradle.openapi.generator.model.name.Name;
import com.github.muehmar.gradle.openapi.generator.settings.JavaModifier;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.generator.settings.TestPojoSettings;
import com.github.muehmar.gradle.openapi.snapshot.SnapshotTest;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import java.util.HashMap;
import org.junit.jupiter.api.Test;

@SnapshotTest
class DiscriminatorValidationMethodGeneratorTest {
  private Expect expect;

  @Test
  @SnapshotName("NoDiscriminator")
  void generate_when_calledWithoutDiscriminator_then_noContent() {
    final Generator<JavaObjectPojo, PojoSettings> generator =
        discriminatorValidationMethodGenerator();
    final Writer writer =
        generator.generate(
            JavaPojos.oneOfPojo(sampleObjectPojo1(), sampleObjectPojo2()),
            defaultTestSettings(),
            javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("DiscriminatorWithoutMapping")
  void generate_when_calledWithDiscriminator_then_correctOutput() {
    final Generator<JavaObjectPojo, PojoSettings> generator =
        discriminatorValidationMethodGenerator();

    final Discriminator noMappingDiscriminator =
        Discriminator.fromPropertyName(
            JavaPojoMembers.requiredString().getName().getOriginalName());

    final JavaObjectPojo pojo =
        JavaPojos.oneOfPojo(
            JavaOneOfCompositions.fromPojosAndDiscriminator(
                NonEmptyList.of(sampleObjectPojo1(), sampleObjectPojo2()), noMappingDiscriminator));

    final Writer writer = generator.generate(pojo, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("IllegalIdentifierPojoDiscriminatorWithoutMapping")
  void generate_when_illegalIdentifierPojo_then_correctOutput() {
    final Generator<JavaObjectPojo, PojoSettings> generator =
        discriminatorValidationMethodGenerator();

    final Discriminator noMappingDiscriminator =
        Discriminator.fromPropertyName(
            JavaPojoMembers.keywordNameString().getName().getOriginalName());

    final JavaObjectPojo pojo =
        JavaPojos.oneOfPojo(
            JavaOneOfCompositions.fromPojosAndDiscriminator(
                NonEmptyList.of(illegalIdentifierPojo(), illegalIdentifierPojo()),
                noMappingDiscriminator));

    final Writer writer = generator.generate(pojo, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("DiscriminatorWithMapping")
  void generate_when_calledWithDiscriminatorAndMapping_then_correctOutput() {
    final Generator<JavaObjectPojo, PojoSettings> generator =
        discriminatorValidationMethodGenerator();

    final HashMap<String, Name> mapping = new HashMap<>();
    mapping.put("obj1", sampleObjectPojo1().getSchemaName().getOriginalName());
    mapping.put("obj2", sampleObjectPojo2().getSchemaName().getOriginalName());

    final Name propertyName = JavaPojoMembers.requiredString().getName().getOriginalName();
    final Discriminator discriminator =
        Discriminator.fromPropertyNameAndMapping(propertyName, mapping);

    final JavaObjectPojo pojo =
        JavaPojos.oneOfPojo(
            JavaOneOfCompositions.fromPojosAndDiscriminator(
                NonEmptyList.of(sampleObjectPojo1(), sampleObjectPojo2()), discriminator));

    final Writer writer = generator.generate(pojo, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("DiscriminatorWithoutMappingWithProtectedAndDeprecatedValidationSettings")
  void generate_when_protectedAndDeprecatedValidationSettings_then_correctOutput() {
    final Generator<JavaObjectPojo, PojoSettings> generator =
        discriminatorValidationMethodGenerator();

    final HashMap<String, Name> mapping = new HashMap<>();
    mapping.put("obj1", sampleObjectPojo1().getSchemaName().getOriginalName());
    mapping.put("obj2", sampleObjectPojo2().getSchemaName().getOriginalName());

    final Name propertyName = JavaPojoMembers.requiredString().getName().getOriginalName();
    final Discriminator discriminator =
        Discriminator.fromPropertyNameAndMapping(propertyName, mapping);

    final JavaObjectPojo pojo =
        JavaPojos.oneOfPojo(
            JavaOneOfCompositions.fromPojosAndDiscriminator(
                NonEmptyList.of(sampleObjectPojo1(), sampleObjectPojo2()), discriminator));

    final Writer writer =
        generator.generate(
            pojo,
            defaultTestSettings()
                .withValidationMethods(
                    TestPojoSettings.defaultValidationMethods()
                        .withModifier(JavaModifier.PROTECTED)
                        .withDeprecatedAnnotation(true)),
            javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }
}
