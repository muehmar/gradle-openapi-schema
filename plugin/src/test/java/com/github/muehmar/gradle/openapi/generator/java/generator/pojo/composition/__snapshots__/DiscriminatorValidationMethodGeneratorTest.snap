DiscriminatorWithMapping=[
com.fasterxml.jackson.annotation.JsonIgnore
javax.validation.constraints.AssertTrue

@AssertTrue(message = "Not valid against the schema described by the discriminator")
@JsonIgnore
private boolean isValidAgainstTheCorrectSchema() {
  if (stringVal == null) {
    return false;
  }
  switch(stringVal) {
    case "obj1": return isValidAgainstSampleObjectPojo1Dto();
    case "obj2": return isValidAgainstSampleObjectPojo2Dto();
  }
  return false;
}
]


DiscriminatorWithoutMapping=[
com.fasterxml.jackson.annotation.JsonIgnore
javax.validation.constraints.AssertTrue

@AssertTrue(message = "Not valid against the schema described by the discriminator")
@JsonIgnore
private boolean isValidAgainstTheCorrectSchema() {
  if (stringVal == null) {
    return false;
  }
  switch(stringVal) {
    case "SampleObjectPojo1": return isValidAgainstSampleObjectPojo1Dto();
    case "SampleObjectPojo2": return isValidAgainstSampleObjectPojo2Dto();
  }
  return false;
}
]


DiscriminatorWithoutMappingWithProtectedAndDeprecatedValidationSettings=[
com.fasterxml.jackson.annotation.JsonIgnore
javax.validation.constraints.AssertTrue

/**
 * @deprecated This method is intended to be used only by frameworks for validation
 * or serialisation.
 */
@AssertTrue(message = "Not valid against the schema described by the discriminator")
@Deprecated
@JsonIgnore
protected boolean isValidAgainstTheCorrectSchema() {
  if (stringVal == null) {
    return false;
  }
  switch(stringVal) {
    case "obj1": return isValidAgainstSampleObjectPojo1Dto();
    case "obj2": return isValidAgainstSampleObjectPojo2Dto();
  }
  return false;
}
]


IllegalIdentifierPojoDiscriminatorWithoutMapping=[
com.fasterxml.jackson.annotation.JsonIgnore
javax.validation.constraints.AssertTrue

@AssertTrue(message = "Not valid against the schema described by the discriminator")
@JsonIgnore
private boolean isValidAgainstTheCorrectSchema() {
  if (switch_ == null) {
    return false;
  }
  switch(switch_) {
    case "Illegal?Identifier": return isValidAgainstIllegal_IdentifierDto();
    case "Illegal?Identifier": return isValidAgainstIllegal_IdentifierDto();
  }
  return false;
}
]


NoDiscriminator=[
.
.

]